Storing a function pointer in a const usize works on nightly

Some people were of the opinion that this shouldn't work, but it currently does. [playground](https://play.rust-lang.org/?gist=d74f852443411fe590122abd83870770&version=nightly&mode=release):

```rust
#![feature(atomic_integers, const_fn)]
use std::sync::atomic::AtomicUsize;
use std::sync::atomic::Ordering;

const unsafe fn transmute(x: *mut u8) -> usize {
    union T {
        a: *mut u8,
        b: usize
    }
    T { a: x }.b
}

const BAR: *mut u8 = ((|| 3) as fn() -> i32) as *mut u8;
const FOO: AtomicUsize = AtomicUsize::new(unsafe { transmute(BAR) });
// static FOO: AtomicUsize = AtomicUsize::new(unsafe { transmute(BAR) }); // ALSO OK
// const BAZ: AtomicUsize = AtomicUsize::new(BAR as usize); ERRORs

fn main() {
    let l = FOO.load(Ordering::Relaxed);
    let l: fn() -> i32 = unsafe { std::mem::transmute(l) };
    assert_eq!(l(), 3);
}
```

cc @rkruppe @eddyb @oli-obk

[It seems](https://play.rust-lang.org/?version=nightly&mode=release&edition=2015&gist=e12ae9a840ee31ab41a7d9b6f8767fdb) this is no longer allowed by any means:

```
#![feature(const_raw_ptr_to_usize_cast)]

const BAR: *mut () = ((|| 3) as fn() -> i32) as *mut ();
pub const FOO: usize = unsafe { BAR as usize };

fn main() {}
```
```
   Compiling playground v0.0.1 (/playground)
error[E0080]: it is undefined behavior to use this value
 --> src/main.rs:4:1
  |
4 | pub const FOO: usize = unsafe { BAR as usize };
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type validation failed: encountered a pointer, but expected initialized plain (non-pointer) bytes
  |
  = note: The rules on what exactly is undefined behavior aren't clear, so this check might be overzealous. Please open an issue on the rust compiler repository if you believe it should not be considered undefined behavior

error: aborting due to previous error

For more information about this error, try `rustc --explain E0080`.
error: Could not compile `playground`.

To learn more, run the command again with --verbose.
```

--- a/src/test/ui/consts/issue-51559.rs
+++ b/src/test/ui/consts/issue-51559.rs
@@ -1,0 +1,7 @@
+#![feature(const_raw_ptr_to_usize_cast)]
+
+const BAR: *mut () = ((|| 3) as fn() -> i32) as *mut ();
+pub const FOO: usize = unsafe { BAR as usize };
+//~^ ERROR it is undefined behavior to use this value
+
+fn main() {}

--- a/src/test/ui/consts/issue-51559.stderr
+++ b/src/test/ui/consts/issue-51559.stderr
@@ -1,0 +1,11 @@
+error[E0080]: it is undefined behavior to use this value
+  --> $DIR/issue-51559.rs:4:1
+   |
+LL | pub const FOO: usize = unsafe { BAR as usize };
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type validation failed: encountered a pointer, but expected initialized plain (non-pointer) bytes
+   |
+   = note: The rules on what exactly is undefined behavior aren't clear, so this check might be overzealous. Please open an issue on the rust compiler repository if you believe it should not be considered undefined behavior
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0080`.

